brontide/noise.go

```diff
@@ -750,31 +756,30 @@ func (b *Machine) ReadMessage(r io.Reader) ([]byte, error) {
 		return nil, err
 	}
 
-	// If this is the first message being read, take a read buffer from the
-	// buffer pool. This is delayed until this point to avoid allocating
-	// read buffers until after the peer has successfully completed the
-	// handshake, and is ready to begin sending lnwire messages.
-	if b.nextCipherText == nil {
-		b.nextCipherText = readBufferPool.Take()
-		runtime.SetFinalizer(b, freeReadBuffer)
-	}
-
-	// Next, using the length read from the packet header, read the
-	// encrypted packet itself.
+	// Compute the packet length that we will need to read off the wire.
 	pktLen := uint32(binary.BigEndian.Uint16(pktLenBytes)) + macSize
-	if _, err := io.ReadFull(r, b.nextCipherText[:pktLen]); err != nil {
+
+	var rawBytes []byte
+	err = readPool.Submit(func(buf *buffer.Read) error {
```

Comment:
Isn't it the case that this `io.Reader` above has already been obtained from the pool? At least in our current use case where this is _always_ called from within `peer.go`. So as is, this will two two items from the pool rather than a single item.

Reply:
Ok, I see the diff has been updated now to remove the pool usage from the peer and instead have it at the brontide layer. If we move it back into the peer, then we don't need to burden the brontide layer with things like read timeous. For example, as is now, I can no longer ask the `brontide.Machine` to read a message from a file on-disk that really just contains a set of regular packed brontide messages.

Reply:
The reader isn't obtained from the pool, it's access to the read buffer that's scheduled. 

> Ok, I see the diff has been updated now to remove the pool usage from the peer and instead have it at the brontide layer.

read pool usage was always in the brontide layer, only write pool usage was in the peer.

this is more of an artifact of the framing protocol, since we can't read/write incrementally to the reader/writer. because of the framing, we have to do
writes:
 1. peer encodes full message (requires write pool)
 2. pass message.bytes() to conn
 3. brontide internally writes full bytes to the wire

reads:
 1. brontide reads and decrypt full message (requires read pool)
 2. retrieve message bytes out of conn
 3. peer decodes full message

step 2 in both cases is where the abstraction should logically happen, but the framing prevents that

Reply:
> For example, as is now, I can no longer ask the `brontide.Machine` to read a message from a file on-disk that really just contains a set of regular packed brontide messages.

we could instead have the type assertions suround the calls to `SetReadDeadline`, instead of erroring if the reader does not match?

Reply:
in fact, i think this is actually necessary for the unit tests to pass